warehouse/manage/views.py

```diff
@@ -1059,6 +1070,232 @@ def toggle_2fa_requirement(self):
+            owner_roles = (
+                self.request.db.query(User.id)
+                .join(Role.user)
+                .filter(Role.role_name == "Owner", Role.project == self.project)
+                .subquery()
+            )
+            owner_users = (
+                self.request.db.query(User)
+                .join(owner_roles, User.id == owner_roles.c.id)
+                .all()
+            )
+            for user in owner_users:
+                send_oidc_provider_added_email(
+                    self.request, user, project_name=self.project.name
+                )
+
+            self.project.oidc_providers.append(provider)
+
+            self.project.record_event(
+                tag="project:oidc:provider-added",
+                ip_address=self.request.remote_addr,
+                additional={
+                    "provider": provider.provider_name,
+                    "id": str(provider.id),
+                    "specifier": str(provider),
+                },
+            )
+
+            self.request.session.flash(
+                f"Added {provider} to {self.project.name}",
+                queue="success",
+            )
+
+        return response
+
+    @view_config(request_method="POST", request_param=DeleteProviderForm.__params__)
+    def delete_oidc_provider(self):
+        if self.request.flags.enabled(AdminFlagValue.DISALLOW_OIDC):
+            self.request.session.flash(
+                (
+                    "OpenID Connect is temporarily disabled. "
+                    "See https://pypi.org/help#admin-intervention for details."
+                ),
+                queue="error",
+            )
+            return self.default_response
+
+        form = DeleteProviderForm(self.request.POST)
+
+        if form.validate():
+            provider = self.request.db.query(OIDCProvider).get(form.provider_id.data)
+
+            # provider will be `None` here if someone manually futzes with the form.
+            if provider is None or provider not in self.project.oidc_providers:
+                self.request.session.flash(
+                    f"Unknown publisher: {provider}",
+                    queue="error",
+                )
+                return self.default_response
+
+            # We only email project owners, since only owners can manage OIDC providers.
+            owner_roles = (
+                self.request.db.query(User.id)
+                .join(Role.user)
+                .filter(Role.role_name == "Owner", Role.project == self.project)
+                .subquery()
+            )
+            owner_users = (
+                self.request.db.query(User)
+                .join(owner_roles, User.id == owner_roles.c.id)
+                .all()
+            )
+            for user in owner_users:
+                send_oidc_provider_removed_email(
+                    self.request, user, project_name=self.project.name
+                )
+
+            self.project.oidc_providers.remove(provider)
```

Comment:
Yeah, I wasn't sure whether it made sense to delete it (or what the deletion semantics should be, since it's a many-many mapping).

I'll add a note.

Reply:
[c2ca980](https://github.com/pypa/warehouse/pull/10753/commits/c2ca980fc4ae3551631d5434212b4b7a0d2f1085)